File: src/lib/exchange-rates.ts
import type { ExchangeRate } from '../types';
import { insertExchangeRates, getExchangeRates } from './db';

const FRANKFURTER_BASE = 'https://api.frankfurter.app';

interface FrankfurterResponse {
  base: string;
  start_date: string;
  end_date: string;
  rates: Record<string, { EUR: number }>;
}

/**
 * Fetch USD→EUR exchange rates for a date range from the Frankfurter API (ECB data).
 * The Frankfurter API returns rates with EUR as one of the target currencies.
 * When from=USD, the rate for EUR means: 1 USD = X EUR.
 */
export async function fetchExchangeRates(startDate: string, endDate: string): Promise<ExchangeRate[]> {
  // Frankfurter API: GET /{start_date}..{end_date}?from=USD&to=EUR
  const url = `${FRANKFURTER_BASE}/${startDate}..${endDate}?from=USD&to=EUR`;
  const response = await fetch(url);

  if (!response.ok) {
    throw new Error(`Failed to fetch exchange rates: ${response.status} ${response.statusText}`);
  }

  const data: FrankfurterResponse = await response.json();
  const rates: ExchangeRate[] = [];

  for (const [date, rateObj] of Object.entries(data.rates)) {
    rates.push({
      date,
      usdToEur: rateObj.EUR,
    });
  }

  // Sort by date
  rates.sort((a, b) => a.date.localeCompare(b.date));
  return rates;
}

/**
 * Ensure we have exchange rates cached for all needed dates.
 * Fetches from the API and stores in SQLite.
 */
export async function ensureRatesForDates(dates: string[]): Promise<Map<string, number>> {
  if (dates.length === 0) return new Map();

  const sorted = [...dates].sort();
  // Extend range by a few days on each side to handle weekends/holidays
  const minDate = shiftDate(sorted[0], -5);
  const maxDate = shiftDate(sorted[sorted.length - 1], 5);

  const existingRates = getExchangeRates();

  // Check if we already have sufficient coverage
  const needsFetch = dates.some((d) => !findRate(d, existingRates));

  if (needsFetch) {
    const fetched = await fetchExchangeRates(minDate, maxDate);
    insertExchangeRates(fetched);
    // Reload from DB
    return getExchangeRates();
  }

  return existingRates;
}

/**
 * Find the exchange rate for a given date.
 * If the exact date is not available (weekend/holiday), walk backward to find the previous business day.
 */
export function findRate(date: string, rates: Map<string, number>): number | null {
  // Try exact date first
  if (rates.has(date)) return rates.get(date)!;

  // Walk backward up to 7 days
  for (let i = 1; i <= 7; i++) {
    const prev = shiftDate(date, -i);
    if (rates.has(prev)) return rates.get(prev)!;
  }

  return null;
}

function shiftDate(dateStr: string, days: number): string {
  const d = new Date(dateStr + 'T00:00:00Z');
  d.setUTCDate(d.getUTCDate() + days);
  return d.toISOString().slice(0, 10);
}

File: src/rsu_tax/exchange_rates.py
"""Fetch USD→EUR exchange rates from the Frankfurter API (ECB data)."""

from __future__ import annotations

from datetime import date, timedelta

import httpx

_FRANKFURTER_BASE = "https://api.frankfurter.app"
_RATE_SANITY_MIN = 0.60
_RATE_SANITY_MAX = 1.15


def _shift_date(d: date, days: int) -> date:
    return d + timedelta(days=days)


def _parse_date(s: str) -> date:
    return date.fromisoformat(s)


async def fetch_rates(start: str, end: str) -> dict[str, float]:
    """
    Fetch daily USD→EUR rates for [start, end] from Frankfurter API.
    Returns a dict mapping 'YYYY-MM-DD' → rate (1 USD = X EUR).
    """
    url = f"{_FRANKFURTER_BASE}/{start}..{end}?from=USD&to=EUR"
    async with httpx.AsyncClient(timeout=15) as client:
        resp = await client.get(url)
        resp.raise_for_status()

    data = resp.json()
    return {
        day: rate_obj["EUR"]
        for day, rate_obj in data.get("rates", {}).items()
    }


async def rates_for_dates(dates: list[str]) -> dict[str, float]:
    """
    Fetch rates for all given dates, extending the range by ±5 days to
    cover weekends and public holidays.
    """
    if not dates:
        return {}

    sorted_dates = sorted(dates)
    start = _shift_date(_parse_date(sorted_dates[0]), -5).isoformat()
    end = _shift_date(_parse_date(sorted_dates[-1]), 5).isoformat()

    return await fetch_rates(start, end)


def find_rate(date_str: str, rates: dict[str, float]) -> float | None:
    """
    Look up the exchange rate for a date.
    Falls back up to 7 days earlier (to handle weekends / public holidays).
    """
    result = find_rate_with_date(date_str, rates)
    return result[0] if result else None


def find_rate_with_date(
    date_str: str, rates: dict[str, float]
) -> tuple[float, str] | None:
    """
    Look up the exchange rate for a date, also returning the actual ECB publication date.
    Falls back up to 7 days earlier (to handle weekends / public holidays).
    Returns (rate, effective_date) or None.
    """
    if date_str in rates:
        return rates[date_str], date_str

    d = _parse_date(date_str)
    for i in range(1, 8):
        prev = _shift_date(d, -i).isoformat()
        if prev in rates:
            return rates[prev], prev

    return None
